### PR TITLE
submission for creating a customised running environment

### PR DESCRIPTION
**Pull request by rod-meaney**
# PR Summary

## PR Checklist

- Do you want this post to be published on a specific date?: Y/N [Specify date]:
- [x] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [x] PR has a meaningful title
- [ ] Includes content related an open issue
  - see [Closing issues using keywords][key].
- [x] This PR is ready to merge and is not **Work in Progress**

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

**Comment by rod-meaney** [comment]
@microsoft-github-policy-service agree

File: Posts/2024/02/creating-a-scalable-customised-running-environment.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+```
+
+What does this do and why does it work?
+
+- At a high level, iterates over the current folder, and runs every .ps1 file as PowerShell
+- `$PSScriptRoot` is the key here, and tells running session, what the location of the current 
+  code is
+
+This means you can create cmd-lets under this structure, and they will automatically load when you
+start up a new PowerShell session.
```
Comment (michaeltlombardi) [inline Posts/2024/02/creating-a-scalable-customised-running-environment.md:61]:
```suggestion
- At a high level, it iterates over the current folder, and runs every `.ps1` file as PowerShell.
- `$PSScriptRoot` is the key here, and tells running session, what the location of the current 
  code is.

This means you can create cmdlets under this structure, and they will automatically load when you
start up a new PowerShell session.
```

File: Posts/2024/02/creating-a-scalable-customised-running-environment.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+Profile.ps1. You will either find it or need to create it in
+
+- PowerShell 5 and lower : `$HOME\Documents\WindowsPowerShell\Profile.ps1`
+- PowerShell 7 : `$HOME\Documents\PowerShell\Profile.ps1`
+- For detailed information, see [About Profiles][05] (Chose 7 or 5)
+
+So this file runs at the start of every session that is opened on your machine. I have included
+both 5 and 7, as in a lot of corporate environments, 5 is all that is available, and often people
+don't have access to modify their environment. With some simple code we can ensure our modules will
+open. Add this into your Profile.ps1:
```
Comment (michaeltlombardi) [inline Posts/2024/02/creating-a-scalable-customised-running-environment.md:75]:
```suggestion
So, the modules are sorted. How do we make sure the modules themselves load? It's all about the
`Profile.ps1`. You will either find it or need to create it in:

- PowerShell 5 and lower - `$HOME\Documents\WindowsPowerShell\Profile.ps1`.
- PowerShell 7 - `$HOME\Documents\PowerShell\Profile.ps1`.
- For detailed information, see [About Profiles][05].

So this file runs at the start of every session that is opened on your machine. I have included
both 5 and 7, as in a lot of corporate environments, 5 is all that is available, and often people
don't have access to modify their environment. With some simple code we can ensure our modules will
open. Add this into your `Profile.ps1`:
```

File: Posts/2024/02/creating-a-scalable-customised-running-environment.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+}
+foreach ($key in $MyModuleDef.Keys){
+    $MyModulePath = $MyModuleDef[$key]
+	$env:PSModulePath = $env:PSModulePath + $([System.IO.Path]::PathSeparator)+$MyModulePath.path
+	$exclude = $MyModulePath.exclude
+	Get-ChildItem -Path $MyModulePath.path -Directory -Exclude $exclude | ForEach-Object{
+		Write-Host "Loading Module $($_.Name) in $Key"
+		Import-Module $_.Name
+	}
+}
```
Comment (michaeltlombardi) [inline Posts/2024/02/creating-a-scalable-customised-running-environment.md:98]:
```suggestion
Write-Host "Loading Modules for Day-to-Day use"
$ErrorActionPreference = "Stop" # A safeguard for any errors

$MyModuleDef = @{
    Utilities = @{
        Path    = "C:\work\git-personal\ps-community-blog\my-utilities"
        Exclude = @(".git")
    }
    Support = @{
        Path    = "C:\work\git-personal\ps-community-blog\my-support"
        Exclude = @(".git")
    }
}

foreach ($key in $MyModuleDef.Keys) {
    $module  = $MyModuleDef[$key]
    $exclude = $module.Exclude

    $env:PSModulePath = @(
        $env:PSModulePath
        $module.Path
    ) -join [System.IO.Path]::PathSeparator

    Get-ChildItem -Path $module.Path -Directory -Exclude $exclude |
        ForEach-Object {
            Write-Host "Loading Module $($_.Name) in $Key"
            Import-Module $_.Name
        }
}
```

File: Posts/2024/02/creating-a-scalable-customised-running-environment.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+What does this do and why does it work?
+
+- At a high level, defines your module groupings, then loads your modules into the PowerShell 
+  session
+- `$MyModuleDef` contains the reference to your module groupings, to make sure all the sub folders 
+  are loaded as modules
+- `exclude` is very important. You may load the code directly of your code base, so ignoring those
+  as modules is important. I have also put dll's in folders in module groupings, and ignoring these
+  is important as well.
```
Comment (michaeltlombardi) [inline Posts/2024/02/creating-a-scalable-customised-running-environment.md:109]:
```suggestion
- At a high level, it defines your module groupings, then loads your modules into the PowerShell
  session.
- `$MyModuleDef` contains the reference to your module groupings, to make sure all the sub folders
  are loaded as modules.
- `Exclude` is very important. You may load the code directly of your code base, so ignoring those
  as modules is important. I have also put DLL's in folders in module groupings, and ignoring these
  is important as well.
```

File: Posts/2024/02/creating-a-scalable-customised-running-environment.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+Now, every time you open any PowerShell session on your machine, all your local cmd-lets will be 
+there, ready to use with all the wonderful functionality you have created.
```
Comment (michaeltlombardi) [inline Posts/2024/02/creating-a-scalable-customised-running-environment.md:112]:
```suggestion
Now, every time you open any PowerShell session on your machine, all your local cmdlets will be
there, ready to use with all the wonderful functionality you have created.
```

File: Posts/2024/02/creating-a-scalable-customised-running-environment.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+## Conclusion
+
+Having your own PowerShell cmd-lets at your fingertips with minimal overhead or thinking makes your
+PowerShell experinece so very much more rewarding.  It also makes it easier to do as I like to do
+and start the day with my favourite mantra: 
+
+'Lets break some stuff!'
```
Comment (michaeltlombardi) [inline Posts/2024/02/creating-a-scalable-customised-running-environment.md:120]:
```suggestion
Having your own PowerShell cmdlets at your fingertips with minimal overhead or thinking makes your
PowerShell experinece so very much more rewarding. It also makes it easier to do as I like to do
and start the day with my favourite mantra: 

> Lets break some stuff!
```

**Review by michaeltlombardi — APPROVED** [review]

**Review by rod-meaney — COMMENTED** [review]
thanks, appreciate the feedback and changes

**Pull request merged**